internal/cms/signed.go

```diff
@@ -0,0 +1,316 @@
+	// verify signature
+	algorithm := oid.ToSignatureAlgorithm(
+		signerInfo.DigestAlgorithm.Algorithm,
+		signerInfo.SignatureAlgorithm.Algorithm,
+	)
+	if algorithm == x509.UnknownSignatureAlgorithm {
+		return VerificationError{Message: "unknown signature algorithm"}
+	}
+
+	signed := d.Content
+	if len(signerInfo.SignedAttributes) > 0 {
+		encoded, err := asn1.MarshalWithParams(signerInfo.SignedAttributes, "set")
+		if err != nil {
+			return VerificationError{Message: "invalid signed attributes", Detail: err}
+		}
+		signed = encoded
+	}
+
+	if err := cert.CheckSignature(algorithm, signed, signerInfo.Signature); err != nil {
+		return VerificationError{Detail: err}
+	}
+	return nil
+}
+
+// verifyAttributes verifies the signed attributes.
+//
+// References:
+//   - RFC 5652 5.3 SignerInfo Type
+//   - RFC 5652 5.6 Signature Verification Process
+func (d *ParsedSignedData) verifyAttributes(signerInfo *SignerInfo, chains [][]*x509.Certificate) error {
+	// verify attributes if present
+	if len(signerInfo.SignedAttributes) == 0 {
+		if d.ContentType.Equal(oid.Data) {
+			return nil
+		}
+		// signed attributes MUST be present if the content type of the
+		// EncapsulatedContentInfo value being signed is not id-data.
+		return VerificationError{Message: "missing signed attributes"}
+	}
+
+	var contentType asn1.ObjectIdentifier
+	if err := signerInfo.SignedAttributes.TryGet(oid.ContentType, &contentType); err != nil {
+		return VerificationError{Message: "invalid content type", Detail: err}
+	}
+	if !d.ContentType.Equal(contentType) {
+		return VerificationError{Message: fmt.Sprintf("mismatch content type: found %q in signer info, and %q in signed data", contentType, d.ContentType)}
+	}
+
+	var expectedDigest []byte
+	if err := signerInfo.SignedAttributes.TryGet(oid.MessageDigest, &expectedDigest); err != nil {
+		return VerificationError{Message: "invalid message digest", Detail: err}
+	}
+	hash, ok := oid.ToHash(signerInfo.DigestAlgorithm.Algorithm)
+	if !ok {
+		return VerificationError{Message: "unsupported digest algorithm"}
+	}
+	actualDigest, err := hashutil.ComputeHash(hash, d.Content)
+	if err != nil {
+		return VerificationError{Message: "hash failure", Detail: err}
+	}
+	if !bytes.Equal(expectedDigest, actualDigest) {
+		return VerificationError{Message: "mismatch message digest"}
+	}
+
+	// sanity check on signing time
+	var signingTime time.Time
+	if err := signerInfo.SignedAttributes.TryGet(oid.SigningTime, &signingTime); err != nil {
+		if errors.Is(err, ErrAttributeNotFound) {
+			return nil
+		}
+		return VerificationError{Message: "invalid signing time", Detail: err}
+	}
+
+	// verify signing time is within the validity period of all certificates
+	// in the chain. As long as one chain is valid, the signature is valid.
+	for _, chain := range chains {
+		if isSigningTimeValid(chain, signingTime) {
+			return nil
+		}
+	}
+
+	return VerificationError{Message: "signature was generated outside certificate's validity period"}
```

Comment:
In this case, let's display the signing time in the error message.

Reply:
Updated.